source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -339,6 +339,9 @@ FilterDataStatus Filter::onData(ProcessorState& state, Buffer::Instance& data, b
     if (state.partialBodyProcessed()) {
       // We already sent and received the buffer, so everything else just falls through.
       ENVOY_LOG(trace, "Partial buffer limit reached");
+      // Make sure that we do not accidentally try to modify the headers before
+      // we continue, which will result in them possibly being sent.
+      state.setHeaders(nullptr);
```

Comment:
For my understanding - do we end up restoring the headers in the processingState somewhere so that we are able to process the headers once we are ready to send them?

Reply:
What actually happened here is that the filter, in [en|de]codeHeaders, stashed away a pointer to the headers if it was going to be in buffered or partial buffered mode so that we can modify them in response to the message body. This only works because in these two cases we have paused processing and we know that envoy will not deliver the headers to the client yet, and because Envoy doesn't move or modify the target of that pointer.

I don't really LIKE doing it this way at all, and I feel like I'm relying too much on the current behavior of the proxy. I would much prefer if Envoy would let me use something reference-counted (like a shared_ptr) to refer to the headers object, but I believe that would require a change to the whole filter mechanism.